license_manager/proto/license_manager.proto

```diff
@@ -0,0 +1,171 @@
+  // invocation, though future implementations may allow for allocations of
+  // multiple and of different types.
+  repeated License license = 1; // required
+
+  // Owning entity issuing the allocation request. Used for logging purposes
+  // only. This could be the name of the user or system issuing the request.
+  string owner = 2; // required
+
+  // Build tag of the allocation request. This tag does not need to be unique
+  // across multiple AllocateRequests, and may be used to associate multiple
+  // such requests with one higher-level task. Typically the Bazel build
+  // invocation ID is used here.
+  string build_tag = 3; // required
+
+  // If this is not the first AllocateRequest message sent for this
+  // invocation, then this is the invocation_id in a previous AllocateResponse
+  // sent, so that this invocation can queue properly. Otherwise, this field
+  // should be left blank.
+  string invocation_id = 4;
+}
+
+message AllocateResponse {
+  oneof response_type {
+    // Returned when allocation is successful; the invocation should be able
+    // to proceed while simultaneously calling Refresh().
+    LicenseAllocated license_allocated = 1;
+
+    // Returned when allocation is unsuccessful due to license contention.
+    // The invocation should continue to poll by calling Allocate() passing
+    // the invocation_id returned in this message.
+    Queued queued = 2;
+  }
+}
+
+message Queued {
+  // Opaque identifier for this invocation, determined by the server. This ID
+  // associates the invocation with a specific spot in the queue; it should be
+  // used in subsequent AllocateRequest messages or the invocation will be
+  // placed at the back of the queue.
+  string invocation_id = 1;
+
+  // Time at which client should issue its next AllocateRequest. The client
+  // should issue its next poll after this time; if it fails to poll for
+  // significantly longer (>5s) it may be moved to the back of the queue.
+  google.protobuf.Timestamp next_poll_time = 2;
+}
+
+message LicenseAllocated {
+  // Opaque identifier for this invocation, determined by the server. This ID
+  // should be used by the client in subsequent RefreshRequests for this
+  // license.
+  string invocation_id = 1;
+
+  // Time at which the allocated license should be refreshed. The client should
+  // issue a RefreshRequest for this invocation_id after this time; if it fails
+  // to refresh for significantly longer (>5s) its allocation may expire.
+  google.protobuf.Timestamp license_refresh_time = 2;
+}
+
+message RefreshRequest {
+  // Opaque identifier for this invocation, as returned by the server in the
+  // AllocateResponse that allocated this license.
+  string invocation_id = 1; // required
+
+  // If true, the invocation is indicating that it no longer requires the
+  // allocated license, and the server should deallocate it. The resulting
+  // RefreshResponse can be expected to have `allocated` == false.
+  bool release = 2;
```

Comment:
I would probably recommend a different call, Release() with RefreshRequest as parameter, if only for clarity of the API.

Reply:
> what if the license the client is attempting to allocate does not exist?
Probably will return `NOT_FOUND` as opposed to `RESOURCE_EXHAUSTED` in the all-out-of-licenses case, to indicate that this case is not retryable. This does require docs, and I can add them as the implementation comes in.

> or if a client refreshes with an invalid invocation id
This probably requires more thought; I'm thinking to support server restarts, the server has a startup period (related to the queue/allocation refresh times) where new and existing queue requests with unknown `invocation_id`s are re-queued, refresh requests for allocations are "adopted" (by decrementing the corresponding counter, in an effort to not have to cancel running invocations) and no new allocations are issued from the queue until the end of the startup period. After the startup period, allocation refresh requests with an unknown ID could either:

* return a successful response with `allocated == false` telling the client to immediately kill the invocation (it doesn't have a license)
* return an error response (maybe `FAILED_PRECONDITION`) and clients should immediately kill the invocation.

I'm tempted to lean towards the latter because returning an error has a higher chance of being logged and therefore alert us more readily to this state that shouldn't happen.

Similarly, for queue refreshes with an unknown ID, we could:

* re-queue the request at the end, successfully
* return an error response (`FAILED_PRECONDITION`)

Leaning towards the former in this case to simplify the client execution - if the client is waiting in the queue and then suddenly its place in the queue is "forgotten", the client would need logic to detect this error and re-queue, vs. the server auto-requeueing the client.

WDYT?

Reply:
> Release() with RefreshRequest as parameter

Are you suggesting to rename `Refresh` to `Release`, or add a separate call named `Release`, or something else?

Reply:
WDYT, first message: +100, well thought, having support for rolling restarts while builds are in progress sounds great. Minor additions:
- client should be written so that if the server goes down, builds in progress keep moving forward.
- I would argue that allocated = true/false is no longer necessary? in the RefreshResponse message? We can just use error values? Refresh successful -> all good. Refresh returning error -> either server is down, or license is no longer allocated, or ... it's part of the error handling code, checking different error status?
- client ... we need to decide what to do if server is down at startup. Assume there are licenses, and move forward? or assume there are no licenses, and wait? probably a flag, but let's discuss this later?

Re Release: I was suggesting having a different Release method, rather than Refresh with release = true. Don't see any big advantage in having a bool flag, separate Release method likely makes for a more straightforward API?

Reply:
> I would argue that allocated = true/false is no longer necessary?

Removed.

>  Assume there are licenses, and move forward? or assume there are no licenses, and wait? probably a flag, but let's discuss this later

Agreed - probably the latter by default, with an override for emergencies.

Reply:
> Re Release: I was suggesting having a different Release method

Done. Thanks!